README
Dyphal, the Dynamic Photo Album
===============================

Dyphal is a web photo album designed to allow photographers to show off their 
work without needing to turn to commercial providers that inject advertising 
and may do ethically dubious things with the photos and information that they 
host.

An album is displayed as a list of photos; each photo can be opened to show a 
larger view with metadata.  The photo's size scales to best fit the browser 
window.  An album can be navigated by mouse, touch, or keyboard.

An album can be hosted on any web document server; no server-side scripts or 
databases are required.  All necessary files are generated from original 
photographs and their metadata up front and stored on the server.  The album 
and its metadata files are designed to require minimal storage space and 
minimal network traffic. 

There is a demo album at <https://ciphertext.info/software/dyphal/#/demo>.


Design principles
-----------------

Dyphal was designed to satisfy the following goals:

1.  The photo is central.  It must be the most prominent element of the page. 
    It must be entirely visible within the browser window; scrolling is not 
    allowed.  Do not up-scale the photo beyond its actual size.  Do down-scale 
    the photo to fit the browser window if necessary.
2.  Metadata is important.  Information about what is in a photo, where and 
    when it was taken, and photographic properties such as exposure and 
    aperture must be visible for each photo.
3.  The album must not break the web.  Browsers' Forward and Back buttons must 
    work.  Bookmarks and saved links to photos and albums must work.  The URI 
    in the address bar must represent the current album or photo. 
4.  The album must work using a document server as the back end.
5.  The album must be as small as possible with as many elements as possible 
    shared between pages.  
6.  The album must use modern web standards with a minimum of cruft to support 
    old browsers.  It must follow security best practices and support CSP.
7.  All features of the album must work properly on recent versions of Chrome 
    and Firefox.  There should be no major problems in minor browsers such as 
    Opera or Edge.  
8.  All content must be in HTML or be inserted at run-time.  All styling must 
    be in CSS or inserted at run-time.
9.  Album styles should be changeable without needing to change scripts.


Generating Dyphal albums
------------------------

Dyphal albums are created using DyphalGenerator: 

1.  Use the **Add Photos** button to import photo files or gThumb 3 catalogs. 
    Photos can be re-ordered by dragging.
2.  Choose the caption and property fields to display for each photo.  The 
    fields available depend on what metadata tags are stored in the selected 
    photos.  By default, only the fields available in *all* selected photos are 
    listed; check the **Show all** box to list all fields that are available in 
    *any* selected photo.  Fields can be re-ordered by dragging.
3.  Set a title, description and footer for the album.  The title and footer 
    will appear on photo pages as well as the album page.  No HTML markup is 
    allowed.
4.  Choose the size for photos in the album.  Larger sizes will show more 
    detail, but require more space and time to load.
5.  Use the **Generate Album** button to save the album and generate the 
    down-scaled photos, thumbnails, and metadata files that it requires.
6.  If you haven't already done so, use the **Install Album Template** button 
    to install the Dyphal web template into the album directory.

If you generated the album "Vacation" in a directory served as 
http://example.com/photos and installed the web template to the same directory, 
you can view the album at http://example.com/photos/#/Vacation.

To edit an album in DyphalGenerator, open the ".dyphal" file for the album.

DyphalGenerator does not currently have a metadata editor.  Instead, it pulls 
metadata fields from embedded photo tags.  The following tags are recognized 
as "properties" (using exiftool's naming scheme): 

*   Composite:Aperture
*   Composite:DigitalZoom
*   Composite:DriveMode
*   Composite:FlashType
*   Composite:FOV
*   Composite:FocalLength35efl
*   Composite:HyperfocalDistance
*   Composite:ImageSize
*   Composite:Lens35efl
*   Composite:LensID
*   Composite:LightValue
*   Composite:ScaleFactor35efl
*   Composite:ShootingMode
*   Composite:ShutterSpeed
*   EXIF:DateTimeOriginal
*   EXIF:ExposureCompensation
*   EXIF:ExposureMode
*   EXIF:Flash
*   EXIF:FocalLength
*   EXIF:ISO
*   EXIF:Make
*   EXIF:Model
*   EXIF:Orientation
*   File:FileSize
*   File:FileType
*   MakerNotes:MacroMode
*   MakerNotes:Rotation

Only three fields are recognized as "captions": Description, Location, and 
Date.  They are synthesized from various tags that are commonly used to store 
such information (again, using exiftool's naming scheme):

*   Description
    *   XMP:Description
    *   IPTC:Caption-Abstract
    *   EXIF:UserComment
*   Location
    *   XMP:Location
    *   IPTC:ContentLocationName
*   Date:
    *   XMP:DateTimeOriginal
    *   Composite:DateTimeCreated
    *   EXIF:DateTimeOriginal

Recent versions of gThumb use these fields for metadata; there are probably 
other photo managers that do as well.  Older versions of gThumb stored metadata 
only in external XML files; the tool "gthumb-comment-update" can be used to 
import metadata from the XML files into embedded tags.


Serving Dyphal albums
---------------------

Nothing special is required to serve Dyphal albums, just a basic web document 
server such as Apache or lighttpd.  For improved security, server operators are 
encouraged to serve Dyphal albums over HTTPS with the following headers:

*   `Content-Security-Policy: default-src 'none'; script-src 'self'; style-src 
    'self'; img-src 'self'; connect-src 'self';`
*   `X-Frame-Options: DENY`
*   `Strict-Transport-Security: max-age=31536000`

Since Dyphal uses asynchronous queries, it won't work from file:// URIs.  

The ".dyphal" files created by DyphalGeneraor should **not** be served with the 
web page.  These files are intended to allow DyphalGenerator to re-open and 
edit albums.


Technical details
-----------------

An album is represented by a single web page, index.html.  JavaScript running 
within that page determines what to display based on hash parameters.  The 
first parameter, which must always be present, identifies the album.  This 
value is the name of the JSON file that describes the album with the ".json" 
suffix omitted.  If your directory structure looks like this:

        album.css
        common.css
        dyphal.js
        index.html
        photo.css
        vacation.json
        photos/photo1.jpeg
        metadata/photo1.jpeg.json
        ...

then the path to load the "vacation" album is 

        /#/vacation

This will load **album view**, which shows the title and description of the 
album as a whole alongside thumbnails for every photo in the album.  Clicking 
on any one of the thumbnails will open **photo view** for that photo, 
containing a larger view of the selected photo and its metadata, along with 
controls to move to the previous or next photos in the album, or to return to 
album view.  The photo will re-scale to the largest size that fits in your 
browser window without hiding the header, footer, or metadata or going beyond 
its physical dimensions.  If the photo is not shown at its full size, clicking 
on it will open **overlay view**  to show the photo at the largest size that 
can fit in the window, potentially hiding the header, footer, and metadata. 
There is also a **help view** that can be opened from either album or photo 
view using the question mark icon.

When loaded in a window smaller than 750 CSS-pixels wide or tall, the album 
will use **compact layout** that omits the metadata and footer from photo 
view in order to show the photo at the largest size possible.  In compact 
layout, overlay view shows the metadata and footer rather than another view of 
the photo.  The album is also displayed full-screen in browsers that support 
this feature if the display is less than 750 CSS-pixels wide or tall (though, 
for technical reasons, the user needs to interact with the page before it can 
enter full-screen mode).

Each photo view has its own URI, consisting of the URI to the album followed by 
the index number of the photo (indices start from 1).  For example, the path 
for the third photo in the "Vacation" album might be

        /#/vacation/3

The photo and help overlays do not have independent URIs.

Since only one HTML document is used for all views, its differences in 
appearance are implemented by switching styles and, to a lesser extent, by 
adding and removing document elements.  Photos, thumbnails, and photo metadata 
are not stored in the HTML; they are loaded from JSON files and their contents 
are inserted or removed from the document when appropriate.  The album JSON 
file (`vacation.json` in the above example) contains a title, footer, and 
description of the album, plus a list of the photos in the album.  Each photo 
has its own JSON file (the photo's name from the album JSON with "`.json`" 
appended), containing the name of the photo file, its dimensions, and its 
metadata.  JSON files, thumbnails, and photos are loaded on demand; the JSON 
and photo for the next photo in the album are pre-cached to reduce load time. 

Appending "`/debug`" to any album or photo URI will load it in debug mode, 
which suppresses some of the styling, outlines various document elements in 
various colours, and logs some information to the browser console.

Everything is implemented in as standards-conformant a manner as possible.  
Browser-specific logic is implemented using feature detection rather than 
browser detection where possible; browser detection is only used to work around 
the buggy touch tracking in Android browsers and the broken scrolling in 
full-screen Internet Explorer because no ways to fix these issues using feature 
detection could be found.

If my JavaScript has idiosyncrasies, it's probably because I'm a C++ programmer 
who learned JavaScript by trial and error.


History
-------

The original version of Dyphal was the BestFit template for gThumb, which I 
wrote in 2005 and became part of gThumb starting with version 2.9.1 in 2006. 
That template was designed to take scale the photo to fit in the browser window 
and show metadata around its edges; superficially, it looked a lot like this 
album.  However, it was limited by the hacks required to support Internet 
Explorer 6 and albums built from that template consisted of a large number of 
repetitive generated HTML files.  

At some point after 2008, gThumb removed the BestFit template without telling 
its author (i.e., me) why.  Since working with the gThumb maintainer in the 
past had been a pain, I opted to not try to get it reinstated.  Instead, I 
dropped support for IE6 (simplifying the code considerably) and started work on 
a version where photos and metadata would be inserted into a single page at 
run-time using JavaScript.  I eventually merged the album and photo pages, 
accidentally re-invented JSON before learning how to use JSON properly, and 
ended up with this.  

The main disadvantage of the new album over the old gThumb-generated one is 
that I can no longer rely on gThumb to select the photos to include, create 
down-sized versions of them, and extract their metadata.  Instead, I wrote my 
own tool, DyphalGenerator.  gThumb, despite a couple annoying mis-features, 
still has the best support for photo metadata of any Linux photo manager that I 
could find, so DyphalGenerator pulls metadata from the places where gThumb 
stores it and pulls lists of photos from gThumb collections.


Security Model
--------------

DyphalGenerator, which generates the server-side data, fills the role of a 
server component.  It is trusted, as are the shell commands that it invokes. 
The image files loaded by DyphalGenerator are *not* trusted and are assumed to 
contain arbitrary malicious content.  The JSON files emitted by DyphalGenerator 
are partially trusted.  The following fields are safe to insert into URIs with 
no additional encoding:

*   `metadataDir`
*   `thumbnail`
*   `path`
*   `name`
*   `photo`

The following fields are restricted to hard-coded values:

*   `orientation`
*   `albumVersion`
*   `captionFields`
*   `propertyFields`
*   `photoResolution`
*   the first element of each `properties` tuple

All other values should be assumed to be attacker-controlled.

To be secure, DyphalGenerator must never emit photo URIs outside of the album 
directory and albums must never execute any scripts or interpret any markup 
found in properties, captions, or other fields.  Additionally, albums must 
never load JSON files or images from outside of the album directory.

See [javascript.html](javascript.html) for information about Dyphal's use of 
JavaScript.

The ".dyphal" files created by DyphalGenerator contain local paths and should 
not be served with the web page.

Why am I so concerned about TOCTOU in DyphalGenerator when there's no privilege 
separation to be attacked?  Practice.


Licence
-------

Dyphal is copyright (c) Rennie deGraaf, 2005-2023.  It is distributed under the 
terms of the GNU General Public Licence, either version 2 or (at your option) 
version 3.  See LICENCE or <http://www.gnu.org/licenses/> for details.


Contact information
-------------------

Rennie deGraaf (rennie-dot-degraaf-at-gmail-dot-com)

The latest version of Dyphal, including full source code, can be obtained at 
<https://ciphertext.info/software/dyphal/>.


Version VERSION, DATE


ChangeLog
2023-03-01  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/dyphal.js: LibreJS support

2023-03-01  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/album.css: left-justify only in small-screen mode to avoid spacing 
      issues.

2023-02-26  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/About.ui, tools/ui-header.py, tools/DyphalGenerator.py: update 
      copyright date to 2023.
    
    * www/dyphal.js, www/index.html: put the album description in the meta 
      "description" tag for the benefit of browsers that use it for bookmark 
      titles.
    
    * tools/photo.py: better support for rotated input images.
      
    * README: minor updates.

2023-02-26  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/DyphalGenerator.py: Rather than sorting by file name, use
      the order from imported gThumb albums.  There was a comment in the
      code that files appeared in arbitrary order, so maybe older
      versions of gThumb 3 behaved differently?  Maybe the correct thing
      to do is to sort by metadata date?


2021-10-03  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/index.html, www/common.css: Fixed the footnote in the help text for
      key bindings.

2021-08-28  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/index.html: added "description" and "referrer" metas

2021-08-15  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/DyphalGenerator.py: fixed an assertion error due to an incorrect 
      Qt5-related check

2021-08-15  Rennie deGraaf <rennie.degraaf@gmail.com>

    * install.sh, tools/DyphalGenerator.py, tools/photo.py: update to PyQt5
      and recent versions of Pandoc.  NOT FULLY TESTED!

    * tools/photo.py: added HDR setting to photo properties that can be 
      displayed

    * tools/photo.py, tools/util.py, test.sh, 
      test/test-DyplayGenerator-LinuxSafeFile.py, 
      test/test-DyphalGenerator-safe_file_open.py: PARTIAL, INCOMPLETE support 
      for opening files on non-Linux platforms.

2018-12-01  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/photorename.sh: Added support for Panasonic's numbering scheme.

2017-06-05  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/photorename.sh: Added a parameter to add to photo numbers.

2017-06-05  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/photorename.sh: Use five-digit numbers in Gthumb XML files, too.

2017-04-22  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/photorename.sh: Use five-digit numbers in photo names rather than 
      four-digit.

2017-01-21  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/About.ui: New file containing a QT Designer definition for an 
      About dialog.

    * tools/DyphalGenerator.py, tools/DyphalGenerator.ui: Added an "About" 
      button to the main window that pops up a modeless dialog.

    * install.sh: Build and install the new dialog UI.

2017-01-21  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/photo.py, www/dyphal.js: Added an albumVersion field to photo 
      JSON.

    * www/dyphal.js: Corrected an unrelated JSLint warning.

2017-01-21  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/album.py: New file containing album parsing and serialization 
      code for DyphalGenerator.  There are now separate ".dyphal" files 
      containing the information necessary to re-load an album; the album 
      JSON files no longer contain local paths or other data that's not 
      necessary for the web page.  This necessitated a file format version.

    * tools/DyphalGenerator: Use the new album load/save code.

    * install.sh: Install the new file.

    * test.sh, test/test_DyphalGenerator_Album_{load,save}.py, 
      test/test_cases/test_DyphalGenerator_Album_{load,save}/*: Test scripts 
      and data for saving and loading albums.

    * test/album/album.json: Renamed to test/album/album_v1.json.

    * test/album/album_v2.dyphal: Test album file for file format v2.

    * www/dyphal.js: Verify that the album JSON is a supported version.

    * README: Documented the new album save files.

2017-01-07  Rennie deGraaf <rennie.degraaf@gmail.com>

    * install.sh: Fixed the version naming logic.  Apparently that git 
      command always sorts lightweight tags ahead of full tags, regardless 
      of the tag date?

    * install.sh: Changed the installation path for template files to better 
      match everything else on my system.

2017-01-07  Rennie deGraaf <rennie.degraaf@gmail.com>

    * tools/DyphalGenerator.py: Removed the reference to ie8compat.js, since 
      that file was deleted in the previous commit.

2017-01-07  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/dyphal.js: Don't request full-screen mode in MSIE 11, in order to 
      work around bug 2017-01-06_001.

    * README, www/dyphal.js, www/index.html, www/ie8compat.js: Drop support 
      for MSIE 8, as photo view was broken in it (bug 2017-01-07_001) and 
      it's been unmaintained for almost a year.  There are probably still 
      various things in the CSS and JS that were only needed for MSIE 8 that 
      should get cleaned up sometime.

2017-01-07  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/dyphal.js, www/index.html, www/photo.css: Center photos vertically 
      in compact layout.

2017-01-06  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/index.html: Fixed an error in the help text and added some more 
      details on key bindings.

2017-01-02  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/common.css: Remove the paragraph margins from footerContent to 
      ensure that the footer doesn't overflow the window in compact layout 
      if the font takes more than 14 pixels for a line of text.  This should 
      fix bug 2017-01-02_002.

2017-01-02  Rennie deGraaf <rennie.degraaf@gmail.com>

    * www/dyphal.js, www/common.css: Ask browsers to display full-screen on 
      small screens.  This doesn't have quite the same thresholds as compact 

tools/photo.py
"""Classes to represent a photo and its properties in DyphalGenerator.
Copyright (c) Rennie deGraaf, 2005-2021.

This program is free software; you can redistribute it and/or modify it 
under the terms of the GNU General Public License as published by the 
Free Software Foundation; either version 2 of the License, or (at your 
option) version 3.

This program is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU 
General Public License for more details.

You should have received a copy of the GNU General Public License 
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""


import time
import re
import os
import subprocess
import json
import urllib.parse
import math
#import wand.image

from PyQt5 import QtWidgets

from dyphal.util import RefCounted, safe_open_file
from dyphal.album import Album

class PropertyError(Exception):
    """Exception raised if a photo property has an unexpected value."""

    def __init__(self, name, value):
        """Initializes a PropertyError."""
        super().__init__()
        self.prop = name
        self.value = value

    def __str__(self):
        """Returns a printable message for a PropertyError."""
        return "Property '%s' had an unexpected value: %s" % (self.prop, self.value)


class Property(object):
    """A photo property."""

    def __init__(self, name, text, default=None, transform=None):
        """Initializes a property."""
        self.name = name
        self.text = text
        self.default = default
        if None is not transform:
            self.transform = transform
        else:
            self.transform = lambda x: x


def transform_exif_time(timestamp):
    """Converts an EXIF timestamp into something more appropriate for 
    viewing."""
    try:
        parsed_time = time.strptime(timestamp, "%Y:%m:%d %H:%M:%S")
        return time.strftime("%Y-%m-%d %H:%M:%SZ", parsed_time)
    except ValueError:
        raise PropertyError("EXIF time", timestamp)


def format_display_time(timestamp):
    """Convert an IPTC or XMP timestamp into something more appropriate for 
    viewing."""
    try:
        parsed_time = None
        if re.match("^[0-9]{4}:[0-9]{2}:[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}[+-][0-9]{2}:[0-9]{2}$", 
                    timestamp):
            # IPTC and XMP timestamps are stored in the following odd format:
            #     YYYY:mm:dd HH:MM:SS[+-]HH:MM
            # strptime() can't parse these because of the colon in the time zone
            parsed_time = time.strptime(timestamp[:-3] + timestamp[-2:], "%Y:%m:%d %H:%M:%S%z")
        elif re.match("^[0-9]{4}:[0-9]{2}:[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}$", timestamp):
            parsed_time = time.strptime(timestamp, "%Y:%m:%d %H:%M:%S")
        formatted_time = time.strftime("%d %B %Y, %H:%M", parsed_time)

        # Python 3.3's strftime ignores tm_gmtoff when formatting "%z", so I need to do it myself.
        # Use RFC 3339 conventions.
        time_zone = "UTC"
        if None is parsed_time.tm_gmtoff:
            time_zone += "-00:00"
        elif 0 > parsed_time.tm_gmtoff:
            time_zone += time.strftime("-%H:%M", time.gmtime(-parsed_time.tm_gmtoff))
        elif 0 < parsed_time.tm_gmtoff:
            time_zone += time.strftime("+%H:%M", time.gmtime(parsed_time.tm_gmtoff))

        return (formatted_time, time_zone)
    except (ValueError, TypeError):
        raise PropertyError("Display time", timestamp)


class PhotoFile(RefCounted, QtWidgets.QListWidgetItem):
    """A photo to add to the album.

    Attributes:
        properties (dict): The properties that have been extracted from 
                this photo.
        captions (dict): The captions that have been extracted from 
                this photo.
        _fileFullPath (str): The path to the file.  May use "~" to 
                represent the user's home directory.  
                eg, "~/Photos/2013-04-03/img_3201a.jpg"
        _fileName (str): The name to use for this photo in the album.  
                Must be unique.  eg, "img_3201a.jpg"
        _jsonName (str): The name of the JSON file for this photo in 
                the album.
        _thumbName (str): The name of the thumbnail file for this photo 
                in the album.
        _file (varies): An object that protects the file from TOCTTOU.
        _width (int): The photo's width in pixels.
        _height (int): The photo's height in pixels.
        _config (Config): A reference to the global configuration object.
    """

    _recognizedProperties = [
        Property("Composite:Aperture", "Aperture", transform=lambda f: "f/"+str(f)),
        Property("Composite:DigitalZoom", "Digital zoom", default="None"),
        Property("Composite:DriveMode", "Drive mode", default="Normal"),
        Property("Composite:FlashType", "Flash type", default="None"),
        Property("Composite:FOV", "Field of view", transform=lambda s: s+"rees"),
        Property("Composite:FocalLength35efl", "Focal length"),
        Property("Composite:HyperfocalDistance", "Hyperfocal distance"),
        Property("Composite:ImageSize", "Image dimensions", transform=lambda s: s+" pixels"),
        #Property("Composite:ISO", "ISO"),
        Property("Composite:Lens35efl", "Lens"),
        Property("Composite:LensID", "Lens ID"),
        Property("Composite:LightValue", "Light value"),
        Property("Composite:ScaleFactor35efl", "Scale factor"),
        Property("Composite:ShootingMode", "Shooting mode"),
        Property("Composite:ShutterSpeed", "Exposure", transform=lambda s: str(s)+" sec."),
        Property("EXIF:DateTimeOriginal", "Creation time", transform=transform_exif_time),
        Property("EXIF:ExposureCompensation", "Exposure compensation"),
        Property("EXIF:ExposureMode", "Exposure mode"),
        Property("EXIF:Flash", "Flash"),
        Property("EXIF:FocalLength", "Focal length"),
        Property("EXIF:ISO", "ISO"),
        Property("EXIF:Make", "Camera make"),
        Property("EXIF:Model", "Camera model"),
        Property("EXIF:Orientation", "Orientation"),
        Property("File:FileSize", "File size"),
        Property("File:FileType", "File type"),
        Property("MakerNotes:MacroMode", "Macro mode"),
        Property("MakerNotes:Rotation", "Rotation", transform=lambda i: str(i)+" degrees"),
        Property("MakerNotes:HDR", "HDR")
    ]

    def __init__(self, filepath, fileName, config):
        """Initializes a PhotoFile.  Opens the file and extracts 
        properties and captions from it."""
        self._config = config
        self._fileName = fileName
        self._fileFullPath = re.sub("^"+os.path.expanduser("~"), "~", filepath)
        super().__init__("%s (%s)" % (self._fileName, self._fileFullPath))
        self._jsonName = self._fileName + ".json"
        (name, suffix) = os.path.splitext(self._fileName)
        self._thumbName = name + ".thumbnail" + suffix
        self._file = safe_open_file(filepath, fileName, config)
        
        try:
            # gThumb stores IPTC strings as UTF-8, but does not set CodedCharacterSet
            properties_text = subprocess.check_output(
                ["exiftool", "-charset", "iptc=UTF8", "-json", "-a", "-G", "-EXIF:Orientation#", 
                 "-All", self._file.getPath()], 
                timeout=self._config.BG_TIMEOUT, universal_newlines=True, stderr=subprocess.STDOUT)
            properties_obj = json.loads(properties_text)[0]

            # exiftool finds way too many properties to force the user to sift through, so we 
            # extract only a hard-coded list of properties that are likely to be interesting.
            self.properties = {}
            for prop in self._recognizedProperties:
                if prop.name in properties_obj:
                    self.properties[prop.text] = prop.transform(properties_obj[prop.name])
                elif None is not prop.default:
                    self.properties[prop.text] = prop.default

            # Override the file name property because exiftool saw our generated file name
            self.properties["File name"] = os.path.basename(filepath)

            # Get the photo dimensions
            self._width = properties_obj["File:ImageWidth"]
            self._height = properties_obj["File:ImageHeight"]
            # If the image needs to be rotated, swap the width and height
            if 5 <= properties_obj["EXIF:Orientation"] <= 8:
                self._width, self._height = self._height, self._width
                if "Image dimensions" in self.properties:
                    self.properties["Image dimensions"] = \
                                                       "%dx%d pixels" % (self._width, self._height)

            self.captions = {}

            # Get the display date, if one exists
            for tag in ["XMP:DateTimeOriginal", "Composite:DateTimeCreated", 
                        "EXIF:DateTimeOriginal"]:
                if tag in properties_obj:
                    (display_time, time_zone) = format_display_time(properties_obj[tag])
                    self.captions["Date"] = display_time
                    self.properties["Time zone"] = time_zone
                    break

            # Get the location, if one exists
            for tag in ["XMP:Location", "IPTC:ContentLocationName"]:
                if tag in properties_obj:
                    self.captions["Location"] = properties_obj[tag]
                    break

            # Get the description, if one exists
            for tag in ["XMP:Description", "IPTC:Caption-Abstract", "EXIF:UserComment"]:
                if tag in properties_obj:
                    self.captions["Description"] = properties_obj[tag]
                    break

        except:
            # If something failed, make sure to not leave any dangling resources.  Ignore any 
            # failures that this causes.
            try:
                self._dispose()
            except:
                pass
            raise

    def _dispose(self):
        """Close a photo file.
        Overrides RefCounted._dispose()."""
        self._file.dispose()

    def getPath(self):
        """Return the path to the photo file."""
        return self._file.getPath()

    def _rescale(self, pixels):
        """Calculate the optimal width and height for the photo to keep 
        it under the given size."""
        aspect = self._width / self._height

        if pixels >= self._width * self._height:
            return (self._width, self._height)
        else:
            # Need to find the largest w,h such that w*h <= pixels and w/h == aspect.
            # w/h == aspect, so w == h*aspect.  Substutiting that for w, h*h*aspect <= pixels.
            # Rearranging for h gives h <= sqrt(pixels/aspect)
            height = math.sqrt(pixels / aspect)
            width = height * aspect
            return (int(width), int(height))

    def getAlbumJSON(self):
        """Return the information about the photo that's necessary for 
        the album JSON file."""
        props = {}
        props["name"] = urllib.parse.quote(self._fileName)
        props["thumbnail"] = urllib.parse.quote(os.path.join(self._config.THUMBNAIL_DIR, 
                                                             self._thumbName))
        props["orientation"] = "horizontal" if self._width >= self._height else "vertical"
        props["path"] = urllib.parse.quote(self._fileFullPath)
        return props

    def generateJSON(self, out_dir_name, width_base, height_base, captions, properties):
        """Generate the JSON file for the photo."""
        (width, height) = self._rescale(width_base * height_base)

        data = {}
        data["albumVersion"] = Album.CURRENT_VERSION
        data["photo"] = urllib.parse.quote(os.path.join(self._config.PHOTO_DIR, self._fileName))
        data["width"] = str(width)
        data["height"] = str(height)
        data["caption"] = \
            [self.captions[tag] for tag in captions if tag in self.captions]
        data["properties"] = \
            [(tag, self.properties[tag]) for tag in properties if tag in self.properties]
        #print(json.dumps(data, indent=2, sort_keys=True))

        with open(os.path.join(out_dir_name, self._jsonName), "w") as json_file:
            json.dump(data, json_file, sort_keys=True)

    def generatePhoto(self, out_dir_name, width_base, height_base, quality):
        """Generate a scaled-down photo."""
        (width, height) = self._rescale(width_base * height_base)
        # See http://www.imagemagick.org/Usage/resize/
        subprocess.check_call(["convert", self.getPath(), "-resize", "%dx%d>" % (width, height), 
                               "-strip", "-quality", str(quality), "-auto-orient",
                               os.path.join(out_dir_name, self._fileName)], 
                              timeout=self._config.BG_TIMEOUT)
        # Doesn't work reliably -- Wand 0.5.9 may throw the following:
        #   wand.exceptions.CacheError: cache resources exhausted `/tmp/tmpfyoznu59/pf1_20210710_130553.jpeg' @ error/cache.c/OpenPixelCache/4083
        #with wand.image.Image(filename=self.getPath()) as img:
        #    img.auto_orient()
        #    img.strip()
        #    img.resize(width=width, height=height)
        #    img.compression_quality = quality
        #    img.save(filename=os.path.join(out_dir_name, self._fileName))

    def generateThumbnail(self, out_dir_name, width_base, height_base, quality):
        """Generate a thumbnail for the photo."""
        # width_base and height_base assume a horizontal photo.  Swap them for a vertical.
        width, height = (width_base, height_base)
        if self._width < self._height:
            width, height = (height_base, width_base)
        # See http://www.imagemagick.org/Usage/thumbnails/
        subprocess.check_call(["convert", self.getPath(), "-thumbnail", "%dx%d^" % (width, height), 
                               "-gravity", "center", "-extent", "%dx%d" % (width, height), 
                               "-quality", str(quality), "-auto-orient",
                               os.path.join(out_dir_name, self._thumbName)], 
                              timeout=self._config.BG_TIMEOUT)
        # Doesn't work reliably -- Wand 0.5.9 may throw the following:
        #   wand.exceptions.CacheError: cache resources exhausted `/tmp/tmpfyoznu59/pf1_20210710_130553.jpeg' @ error/cache.c/OpenPixelCache/4083
        #with wand.image.Image(filename=self.getPath()) as img:
        #    img.auto_orient()
        #    img.thumbnail(width=width, height=height)
        #    img.gravity = 'center'
        #    img.extent(width=width, height=height)
        #    img.compression_quality = quality
        #    img.save(filename=os.path.join(out_dir_name, self._thumbName))
